src/devices/machine/ibmps2.cpp

```diff
@@ -0,0 +1,1113 @@
+	if(m_sram_size != 0) map(0x0074, 0x0076).rw(FUNC(ps2_mb_device::nvram_r), FUNC(ps2_mb_device::nvram_w));
+	map(0x0080, 0x008f).rw(m_dmac, FUNC(ibm72x7377_device::page8_r), FUNC(ibm72x7377_device::page8_w)).umask16(0xffff);
+	map(0x0090, 0x0090).rw(m_dmac, FUNC(ibm72x7377_device::dma_arbiter_r), FUNC(ibm72x7377_device::dma_arbiter_w));
+	map(0x0091, 0x0091).r(FUNC(ps2_mb_device::card_select_feedback_r)).w(m_dmac, FUNC(ibm72x7377_device::dma_feedback_w));
+	map(0x0092, 0x0092).rw(FUNC(ps2_mb_device::porta_r), FUNC(ps2_mb_device::porta_w));
+	map(0x0094, 0x0094).rw(FUNC(ps2_mb_device::system_board_pos_r), FUNC(ps2_mb_device::system_board_pos_w));
+	map(0x0096, 0x0096).rw(FUNC(ps2_mb_device::adapter_pos_r), FUNC(ps2_mb_device::adapter_pos_w));
+	map(0x00a0, 0x00a1).rw(m_pic8259_slave, FUNC(pic8259_device::read), FUNC(pic8259_device::write)).umask16(0xffff);
+	map(0x00c0, 0x00df).rw(m_dmac, FUNC(ibm72x7377_device::dmac2_r), FUNC(ibm72x7377_device::dmac2_w)).umask16(0x00ff);
+	map(0x0100, 0x0107).rw(FUNC(ps2_mb_device::pos_registers_r), FUNC(ps2_mb_device::pos_registers_w));
+	// 0x1F0 - Fixed Disk
+	map(0x0680, 0x0681).w(FUNC(ps2_mb_device::post_debug_w));
+}
+
+/*************************************************************
+ *
+ * pic8259 configuration
+ *
+ *************************************************************/
+uint8_t ps2_mb_device::get_slave_ack(offs_t offset)
+{
+	if (offset==2) // IRQ = 2
+		return m_pic8259_slave->acknowledge();
+
+	return 0x00;
+}
+
+/*************************************************************************
+ *
+ *      PC Speaker related
+ *
+ *************************************************************************/
+
+void ps2_mb_device::speaker_set_spkrdata(uint8_t data)
+{
+	m_at_spkrdata = data ? 1 : 0;
+	m_speaker->level_w(m_at_spkrdata & m_pit_out2);
+}
+
+
+
+/*************************************************************
+ *
+ * pit8254 configuration
+ *
+ *************************************************************/
+
+WRITE_LINE_MEMBER( ps2_mb_device::pit8254_out2_changed )
+{
+	m_pit_out2 = state ? 1 : 0;
+	m_speaker->level_w(m_at_spkrdata & m_pit_out2);
+}
+
+
+/*************************************************************************
+ *
+ *      PC DMA stuff
+ *
+ *************************************************************************/
+
+
+void ps2_mb_device::write_rtc(offs_t offset, uint8_t data)
+{
+	if (offset==0) {
+		m_nmi_enabled = BIT(data,7);
+		if (!m_nmi_enabled)
+			m_maincpu->set_input_line(INPUT_LINE_NMI, CLEAR_LINE);
+		m_mc146818->write(0,data);
+	}
+	else {
+		m_mc146818->write(offset,data);
+	}
+}
+
+uint8_t ps2_mb_device::nvram_r(offs_t offset)
+{
+	switch(offset)
+	{
+		case 0:
+			return (m_nvram_index & 0x00FF);
+		case 1:
+			return (m_nvram_index & 0x0700) >> 8;
+		case 2:
+			LOG_NVRAM("reading NVRAM at %04X: %02X\n", m_nvram_index, m_sram[m_nvram_index]);
+			return m_sram[m_nvram_index];
+		default:
+			fatalerror("nvram_r out of range");
```

Comment:
You need to provide more context in the message for a fatal error – it doesn’t automatically include the device tag like `logerror` does.